### PR TITLE
[PREVIEW] RDM-2029 Create new UI component off CasePaymentHistoryViewer type. C…

### PR DESCRIPTION
**Pull request by andrewfolga**
…reate payments_url config. Wire up new component in tabs view.

DO NOT MERGE.

### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/RDM-2029

### Change description ###
Enable gateway to proxy payments calls.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/app/shared/palette/base-field/abstract-field-read.component.ts [hunk 1]
```diff
@@ -1,12 +1,16 @@
 import { Input } from '@angular/core';
 import { CaseField } from '../../domain/definition/case-field.model';
 import { PaletteContext } from './palette-context.enum';
+import { ContextMap } from '@hmcts/ccd-case-ui-toolkit';
```
Comment (mario-paniccia) [inline src/app/shared/palette/base-field/abstract-field-read.component.ts]: same as above

File: src/app/shared/palette/payment/case-payment-history-viewer-field.html [hunk 2]
```diff
@@ -0,0 +1 @@
+<cut-case-payment-history-list [baseURL]="getBaseURL()" [caseReference]="getCaseReference()"></cut-case-payment-history-list>
```
Comment (mario-paniccia) [inline src/app/shared/palette/payment/case-payment-history-viewer-field.html]: 👍

File: src/app/shared/palette/payment/case-payment-history-viewer-field.component.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+@Component({
+  selector: 'ccd-case-payment-history-viewer-field',
+  templateUrl: 'case-payment-history-viewer-field.html',
+  styleUrls: ['case-payment-history-viewer-field.scss']
+})
+export class CasePaymentHistoryViewerFieldComponent extends AbstractFieldReadComponent {
+
+  public static readonly PAYMENTS_BASE_URL_KEY = 'payments_url';
+  public static readonly CASE_REFERENCE_KEY = 'case_reference';
```
Comment (mario-paniccia) [inline src/app/shared/palette/payment/case-payment-history-viewer-field.component.ts]: case reference is very generic concept. This key will probably be used by other components soon.
Shall we move it outside of CasePaymentHistoryViewerFieldComponent?

File: src/app/core/core.module.ts [hunk 4]
```diff
@@ -56,7 +56,6 @@ import { ActivityResolver } from './activity/activity.resolver';
     PaginationService,
     RequestOptionsBuilder,
     HttpService,
-    HttpErrorService,
```
Comment (andrewfolga) [inline src/app/core/core.module.ts]: This was listed twice.

File: src/app/shared/palette/order-summary/read-order-summary-field.component.spec.ts [hunk 5]
```diff
@@ -14,7 +14,7 @@ describe('ReadOrderSummaryFieldComponent', () => {
 
   const FIELD_TYPE: FieldType = {
     id: 'PersonOrderSummary',
-    type: 'OrderSummary'
```
Comment (andrewfolga) [inline src/app/shared/palette/order-summary/read-order-summary-field.component.spec.ts]: This is fixing old bug where OrderSummary was considered a new base type. It is a predefined complex type.

File: src/app/shared/palette/payment/case-payment-history-viewer-field.scss [hunk 6]
```diff
@@ -0,0 +1,8 @@
+@import "~govuk_frontend_toolkit/stylesheets/_colours.scss";
+
+.manual-link {
```
Comment (andrewfolga) [inline src/app/shared/palette/payment/case-payment-history-viewer-field.scss]: This should be changed to provided correct styling.

File: src/app/cases/viewer/case-viewer.component.ts [hunk 7]
```diff
@@ -95,6 +100,13 @@ export class CaseViewerComponent implements OnInit, OnDestroy {
     });
   }
 
+  buildContextMap(): ContextMap {
+    let contextMap = new ContextMap();
+    contextMap.set(Constants.CASE_REFERENCE_KEY, this.caseDetails.case_id);
+    contextMap.set(CasePaymentHistoryViewerFieldComponent.PAYMENTS_BASE_URL_KEY, this.appConfig.getPaymentsUrl());
```
Comment (vlaurin) [inline src/app/cases/viewer/case-viewer.component.ts]: The payment URL being unique for the whole app, I don't see the benefit of passing it via the context. `AppConfig` can be injected directly inside the component.

File: src/app/cases/viewer/case-viewer.component.html [hunk 8]
```diff
@@ -38,10 +38,10 @@ <h3 class="heading-medium error-summary-heading" id="edit-case-event_error-summa
                   <th>
                     <div class="case-viewer-label">{{field.label}}</div>
                   </th>
-                  <td><ccd-field-read [caseField]="field"></ccd-field-read></td>
+                  <td><ccd-field-read [caseField]="field" [caseViewContext]="buildContextMap()"></ccd-field-read></td>
```
Comment (vlaurin) [inline src/app/cases/viewer/case-viewer.component.html]: Those are not the only places where `ccd-field-read` is used. Because `caseViewContext` appears to be a mandatory input, it should be added to every place where `ccd-field-read` is used (read only in wizard, check your answers, case list, search results...).
Incidentally, this requires modifying a few areas that are not directly related to the current change.

File: src/app/cases/viewer/case-viewer.component.ts (hunk 7 above)
Comment (andrewfolga) [reply src/app/cases/viewer/case-viewer.component.ts]: This has been changed now.

File: src/app/cases/viewer/case-viewer.component.html (hunk 8 above)
Comment (andrewfolga) [reply src/app/cases/viewer/case-viewer.component.html]: Do you mean mandatory as in being setup inside FieldReadComponent? If so I think it will not be used as it is not defined for those other places (read only in wizard, check your answers, case list, search results...). Therefore not sure this would require changes in other areas as I felt this is like optional and can be skipped meaning it's not being used.

**Comment by andrewfolga** [comment]
Fatih, Mario and I discussed merging the two contexts but though it would be wiser to do it in 2 steps. First PR to introduce a new context and then a follow up PR to merge both.

File: src/app/shared/palette/payment/case-payment-history-viewer-field.component.ts [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+})
+export class CasePaymentHistoryViewerFieldComponent extends AbstractFieldReadComponent {
+
+  public static readonly PAYMENTS_BASE_URL_KEY = 'payments_url';
+
+  payments_url: string;
+  case_reference: string;
+
+  getBaseURL() {
+    return this.caseViewContext.get(CasePaymentHistoryViewerFieldComponent.PAYMENTS_BASE_URL_KEY);
```
Comment (vlaurin) [inline src/app/shared/palette/payment/case-payment-history-viewer-field.component.ts]: there's no benefit retrieving the URL from the `caseViewContext` when it could be retrieved from `appConfig` directly here.

File: src/app/cases/viewer/case-viewer.component.ts (hunk 7 above)
Comment (vlaurin) [reply src/app/cases/viewer/case-viewer.component.ts]: What I suggested was to remove it completely from the `caseViewContext`, and use `appConfig` directly in `CasePaymentHistoryViewerFieldComponent`, see related comment on `case-payment-history-viewer-field.component.ts`

**Comment by vlaurin** [comment]
This PR is still relevant; updated to point to `master`

File: src/app/shared/palette/payment/case-payment-history-viewer-field.scss (hunk 6 above)
Comment (andrewfolga) [reply src/app/shared/palette/payment/case-payment-history-viewer-field.scss]: This was a copy from CUT library. Can we review this is needed?

File: src/public/config.json [hunk 10]
```diff
@@ -12,9 +12,10 @@
   "print_service_url": "http://localhost:3453/print",
   "remote_print_service_url": "https://return-case-doc.dev.ccd.reform.hmcts.net",
   "smart_survey_url": "https://www.smartsurvey.co.uk/s/CCDfeedback/",
-  "activity_url": "http://localhost:3453/activity",
+  "activity_url": "",
```
Comment (andrewfolga) [inline src/public/config.json]: Please revert.

Comment (vlaurin) [reply src/public/config.json]: Done

File: package.json [hunk 11]
```diff
@@ -52,14 +52,15 @@
     "@angular/platform-server": "~6.0.6",
     "@angular/router": "~6.0.6",
     "@angular/upgrade": "~6.0.6",
-    "@hmcts/ccd-case-ui-toolkit": "1.2.1",
+    "@hmcts/ccd-case-ui-toolkit": "^1.2.1",
```
[the remark this answers is not in the document]
Comment (vlaurin) [reply package.json]: @andrewfolga you tell me https://github.com/hmcts/ccd-case-management-web/commit/6d6adaf65e13d79417c0a60108220e55a7c45f21

[the remark this answers is not in the document]
Comment (andrewfolga) [reply package.json]: Yeah I meant we do not need this change :) We should revert it.

File: src/app/shared/constants.model.ts [hunk 12]
```diff
@@ -0,0 +1,3 @@
+export class Constants {
```
[the remark this answers is not in the document]
Comment (andrewfolga) [reply src/app/shared/constants.model.ts]: Haven't checked exactly but similar to above I think we do not need it and should revert/remove it.

[the remark this answers is not in the document]
Comment (vlaurin) [reply src/app/shared/constants.model.ts]: Yep, removing now.

File: package.json (hunk 11 above)
[the remark this answers is not in the document]
Comment (vlaurin) [reply package.json]: Aye aye captain

[the remark this answers is not in the document]
Comment (andrewfolga) [reply package.json]: :)

[the remark this answers is not in the document]
Comment (andrewfolga) [reply package.json]: I keep leaving comments for myself into the future:)

File: src/app/shared/palette/payment/case-payment-history-viewer-field.scss (hunk 6 above)
Comment (vlaurin) [reply src/app/shared/palette/payment/case-payment-history-viewer-field.scss]: I cannot see that class used in Payments web components, I'm removing it.
